File: linked-list/detectlooplinklist.cpp
class Solution
{
    public:
    //Function to check if the linked list has a loop.
    bool detectLoop(Node* head)
    {
        Node *slow= head;// this is a slow fast pointer algorithm.
        Node *fast= head;
        
        while(fast!=NULL && fast->next!=NULL)// cuz fast pointer is moving two nodes ahead thats why we go upto fast->next!=NULL.
        {
            slow =slow->next;     // we put slow and fast pointers the fast pointer is always set two nodes ahead.
            fast = fast->next->next;
            
            if(slow==fast){
                return true;  // if slow pointer meets the fast pointer then the loop is detected.
            }
        }
        return false; // otherwise the there is no loop.
    }
};

File: linked-list/detectlooplinklist.js
var hasCycle = function(head) {
    let slow = head;
    let fast = head;

    // Step 2: Traverse the linked list
    // with the slow and fast pointers
    while (fast !== null && fast.next !== null) {
        // Move slow one step
        slow = slow.next;
        // Move fast two steps
        fast = fast.next.next;

        // Check if slow and fast pointers meet
        if (slow === fast) {
            return true;  // Loop detected
        }
    }

    // If fast reaches the end of the list, there is no loop
    return false;

};
